Common/Securities/Option/StrategyMatcher/OptionPosition.cs

```diff
@@ -0,0 +1,204 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+
+namespace QuantConnect.Securities.Option.StrategyMatcher
+{
+    /// <summary>
+    /// Defines a lightweight structure representing a position in an option contract or underlying.
```

Comment:
`OptionPosition` instances are _very_ transient and should really only remain around for finding a match solution. These shouldn't be kept in a collection and kept up-to-date on each time step. Instead, when we need to find a strategy match solution, we'll initialize a new `OptionPositionCollection` from the existing `SecurityHolding` instances. Another thing is that `SecurityHolding` is a class and they're all strewn about main memory. The way I **want** the `OptionPositionCollection` data structure to look like is to have a single `ImmutableArray<OptionPosition>` and then all the index collections point to an array index. Since `OptionPosition` is a struct, this would provide for awesome data locality to improve cache hits, vs using `SecurityHolding`, practically every time you reference a property you would need to dereference the pointer and find it in main memory somewhere.

If I'm feeling lucky I'll write up a benchmark and rewrite the core of `OptionPositionCollection` to be backed by a single `ImmutableArray<OptionPosition>` and have the indexes point into that array. I'm betting it would be twice as fast.